type-definitions/Immutable.d.ts

```diff
@@ -761,7 +761,9 @@ declare namespace Immutable {
    */
   export function Map<K, V>(collection?: Iterable<[K, V]>): Map<K, V>;
   export function Map<V>(obj: { [key: string]: V }): Map<string, V>;
-  export function Map<K extends string, V>(obj: { [P in K]?: V }): Map<K, V>;
+  export function Map<K extends string | number, V>(
```

Comment:
`string | symbol` ?

Reply:
I think it should probably be `string | number | symbol` as number can be an object index. But in that case `Map('abc')` and `Map([1,2,3])` do not throw an error anymore. 

The result type is something like that : 
> const abc: Map<number | keyof String, string | number | (() => string) | ((pos: number) => string) | ((index: number) => number) | ((...strings: string[]) => string) | ((searchString: string, position?: number | undefined) => number) | ... 36 more ... | (() => IterableIterator<...>)>

Example here : https://www.typescriptlang.org/play?ssl=1&ssc=51&pln=1&pc=57#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwxAGcMAeAaXhAA9DVgj4SYtUBzeAHyYE8BbAEY4IXeKmSCQMADTwAagD4AFDgEArAFzwA3vADaABXit45ALoB+LXPgBfAJRaAslAAOFWYoDcAWABQ-mB4JEzwALzwAMr8QhBKAOQ4ANbx9oHBGAThUTHCCag4mcmp-v4A9GVRABY4yBDAxqiIUs5upPFQ8aIYPK4gOIhMsuKSMArpqCFwRHWZEYQkSrpQWgBMsnpEZloAzLIALFoArHZpAX7TswB0bCAYCZ1nlxAYN3dKRGcV8AACGEQAWhofUwQJgMBwMH8z1et3u8QEJT83z+gOB4AwYIhUIuxGucKUGDOpWRlVRQOoIMxUmxExCUAEYGyC3hDLASPKZP+FKpWMhdKKqBAABUAO44YVVOAIebEe56ACMsnW8B2ZjOQA

Reply:
I'm not sure if that's a bug in Typescript, or a quirk they're forced to endure because of how JS Array exotics are weird, but in that play link I see `const result = test({ a: 2, [s]: 3, 4: 5 })` inferred to return `Map<'a' | typeof s | 4, number>` when I would have expected `Map<'a' | typeof s | '4', number>` (note the string `"4"` not the number `4`)

It also reports that `result.get(4)` is valid but `result.get("4")` is a type error, but you can try this yourself:

```ts
const result = Immutable.Map({ a: 2, 4: 5 })
result.get(4) === undefined
result.get("4") === 5
```

Javascript Object property names are *always* strings

Reply:
This is by far not the first time I've been bitten by this or have seen someone else bitten by this. I searched through TS's issue log and saw some things roughly related but none calling the direct inference issue out, so I opened one:

https://github.com/microsoft/TypeScript/issues/45170